src/emulate.h

```diff
@@ -0,0 +1,174 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#if RV32_HAS(EXT_F)
+#include <math.h>
+#include "soft-float.h"
+
+#if defined(__APPLE__)
+static inline int isinff(float x)
+{
+    return __builtin_fabsf(x) == __builtin_inff();
+}
+static inline int isnanf(float x)
+{
+    return x != x;
+}
+#endif
+#endif /* RV32_HAS(EXT_F) */
+
+#if RV32_HAS(GDBSTUB)
+extern struct target_ops gdbstub_ops;
+#endif
+
+#include "decode.h"
+#include "riscv.h"
+#include "riscv_private.h"
+#include "utils.h"
+
+#define RVOP(insn) bool rvop_##insn(riscv_t *rv UNUSED, rv_insn_t *ir UNUSED)
+typedef bool (*rvop_t)(riscv_t *rv, rv_insn_t *ir);
+RVOP(lui);
```

Comment:
With the X macro technique, the duplication can be eliminated. See `src/emulate.c` for reference.